File: core/pdf_reader.py
"""Extract embedded PDF text in physical page order; no OCR or outline parsing."""

from pypdf import PdfReader

from core.file_reader import Chapter, _clean_text
from i18n import t


class _PasswordRequired(Exception):
    pass


def read_pdf_file(path: str) -> list[Chapter]:
    chapters = []
    with open(path, "rb") as source:
        try:
            reader = PdfReader(source)
            if reader.is_encrypted and not reader.decrypt(""):
                raise _PasswordRequired
            for page_number, page in enumerate(reader.pages, 1):
                text = _clean_text(page.extract_text() or "")
                if text:
                    chapters.append(Chapter(
                        index=len(chapters) + 1,
                        title=f"Page {page_number:03d}",
                        text=text,
                    ))
        except _PasswordRequired as error:
            raise ValueError(t("error.pdf_password")) from error
        except Exception as error:
            raise ValueError(t("error.pdf_invalid")) from error
    if not chapters:
        raise ValueError(t("error.pdf_no_text"))
    return chapters

File: mobile/lib/services/pdf_reader.dart
import 'dart:typed_data';

import 'package:pdfrx/pdfrx.dart';

import '../models/chapter.dart';

class PdfReadException implements Exception {
  const PdfReadException(this.messageKey);
  final String messageKey;
}

/// Reads embedded text in physical page order, without OCR or outline parsing.
Future<List<Chapter>> readPdfSections(Uint8List bytes) async {
  await pdfrxFlutterInitialize();
  PdfDocument? document;
  try {
    document = await PdfDocument.openData(bytes);
    final chapters = <Chapter>[];
    for (final page in document.pages) {
      final raw = await page.loadText();
      if (raw == null) throw const PdfReadException('pdfInvalid');
      final text = raw.fullText
          .replaceAll('\r\n', '\n')
          .replaceAll('\r', '\n')
          .replaceAll(RegExp(r'[ \t]+'), ' ')
          .replaceAll(RegExp(r' ?\n ?'), '\n')
          .replaceAll(RegExp(r'\n{3,}'), '\n\n')
          .trim();
      if (text.isEmpty) continue;
      chapters.add(Chapter(
        index: chapters.length + 1,
        title: 'Page ${page.pageNumber.toString().padLeft(3, '0')}',
        text: text,
      ));
    }
    if (chapters.isEmpty) throw const PdfReadException('pdfNoText');
    return chapters;
  } on PdfPasswordException {
    throw const PdfReadException('pdfPassword');
  } on PdfException {
    throw const PdfReadException('pdfInvalid');
  } finally {
    await document?.dispose();
  }
}
